src/CapitalInvestments/PyomoModels/PySPBase.py

```diff
@@ -0,0 +1,376 @@
+
+  def createModel(self):
+    """
+      This method is used to create pyomo model.
+      @ In, None
+      @ Out, model, pyomo.AbstractModel, abstract pyomo model
+    """
+    model = super().createModel()
+    return model
+
+  def createInstance(self, data):
+    """
+      This method is used to instantiate the pyomo model
+      @ In, data, dict, dictionary to initialize pyomo abstract model
+      @ Out, model, pyomo.instance, instance of pyomo model
+    """
+    model = super().createInstance(data)
+    # Default to disable some optional constraints
+    if len(self.optionalConstraints) > 0 and self.uncertainties is not None:
+      # if 'consistentConstraintI' in self.optionalConstraints:
+      #   model.consistentConstraint.deactivate()
+      if 'consistentConstraintII' in self.optionalConstraints:
+        model.consistentConstraintII.deactivate()
+        logger.debug('Default to deactivate consistent constraint II')
+      for optCon, decision in self.optionalConstraints.items():
+        if optCon == 'consistentConstraintI' and not decision:
+          model.consistentConstraintI.deactivate()
+          logger.debug('Deactivate consistent constraint I')
+        if optCon == 'consistentConstraintII' and decision:
+          model.consistentConstraintII.activate()
+          logger.debug('Activate consistent constraint II')
+    return model
+
+  def createScenarioTreeModel(self):
+    """
+      Construct scenario tree based on abstract scenario tree model for stochastic programming
+      @ In, None
+      @ Out, treeModel, Instance, pyomo scenario tree model
+    """
+    treeModel = CreateAbstractScenarioTreeModel()
+    if pyoVersion:
+      treeModel = treeModel.create_instance()
+    treeModel.Stages.add('FirstStage')
+    treeModel.Stages.add('SecondStage')
+    treeModel.Nodes.add('RootNode')
+    for i in self.scenarios['scenario_name']:
+      leafNode = 'leaf_' + i
+      treeModel.Nodes.add(leafNode)
+      treeModel.Scenarios.add(i)
+    if not pyoVersion:
+      treeModel = treeModel.create_instance()
+    treeModel.NodeStage['RootNode'] = 'FirstStage'
+    treeModel.ConditionalProbability['RootNode'] = 1.0
+    for node in treeModel.Nodes:
+      if node != 'RootNode':
+        treeModel.NodeStage[node] = 'SecondStage'
+        treeModel.Children['RootNode'].add(node)
+        treeModel.Children[node].clear()
+        treeModel.ConditionalProbability[node] = self.scenarios['probabilities'][node.replace('leaf_','')]
+        treeModel.ScenarioLeafNode[node.replace('leaf_','')] = node
+
+    return treeModel
+
+  @abc.abstractmethod
+  def pysp_scenario_tree_model_callback(self):
+    """
+      scenario tree instance creation callback
+      @ In, None
+      @ Out, treeModel, Instance, pyomo scenario tree model for two stage stochastic programming
+    """
+    pass
+
+  def pysp_instance_creation_callback(self, scenario_name, node_names):
```

Comment:
method naming convention is not uniform throughout the class

Reply:
These are call back functions that are required by pyomo. We need to provide exactly the same name